Query Refactoring: Minor consistency cleanups for default values and validation

cbuescher opened the pull request:
Revisited the queries we already refactored, corrected and aligned some of the codebase based on the conventions that we decided to follow. Also including some cosmetic fixes (making members final where possible, avoiding this references outside of setters/getters).
In addition to that this PR changes:
* preventing NPEs in doXContent when rendering out nested queries that are `null`. We now render out empty object (`{ }`) which then gets parser to `null` to be consistent with queries than come only through the rest layer
* prevent adding nested `null` queries to collections of clauses like in BoolQueryBuilder
* add validate() method to all builders (even when empty)

This PR is against the query-refactoring branch.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -43,11 +43,11 @@
+    public static final boolean DISABLE_COORD_DEFAULT = false;
 
-    static final BoolQueryBuilder PROTOTYPE = new BoolQueryBuilder();
+    public static final BoolQueryBuilder PROTOTYPE = new BoolQueryBuilder();
```
this shouldn't become public right?

cbuescher replied:
Must have thought thats a constant for a second, will change that back.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java:
```diff
@@ -297,17 +297,17 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
-        Queries.applyMinimumShouldMatch(booleanQuery, this.minimumShouldMatch);
-        Query query = this.adjustPureNegative ? fixNegativeQueryIfNeeded(booleanQuery) : booleanQuery;
-        if (this.queryName != null) {
-            parseContext.addNamedQuery(this.queryName, query);
```
why did you remove all the `this.` ?

cbuescher replied:
I found the use of it inconsistent in most methods. I left it in the getters/setters where it makes sense to distinguish between method parameter and class member with the same name. I think I heard there's no real style guide around this, but I wanted to avoid having both variants in the same method at least.

javanna replied:
ok

javanna commented on core/src/main/java/org/elasticsearch/index/query/MatchAllQueryParser.java:
```diff
@@ -41,22 +41,24 @@ public MatchAllQueryParser() {
                 }
             }
         }
+        MatchAllQueryBuilder queryBuilder = new MatchAllQueryBuilder();
```
not sure I get the reason behind this change, wasn't it fine before?

cbuescher replied:
It was okay but in most of the other parsers we follow the pattern of first collecting all the values we need and then constructing the new query builder at the end of the method. I think that way it's easier to read and you have all the code that constructs the return query in one place.

javanna replied:
ok

javanna commented:
left a couple of comments

cbuescher commented:
@javanna I went through the builders again and checked for places where we might need to set default values in setters when argument is `null`. Most of the times argument is some primitive int/float/boolean, so `null` in not possible. When argument is a nested query builder, `null` is unfortunately an option we have to accept and check for later when generating the lucene query. In other cases value is really optional.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java:
```diff
@@ -172,15 +172,15 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
 
     @Override
```
I see one thing that looks weird in this class:

```
@Override
    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
        if (positiveQuery == null) {
            throw new IllegalArgumentException("boosting query requires positive query to be set");
        }
        if (negativeQuery == null) {
            throw new IllegalArgumentException("boosting query requires negative query to be set");
        }
        ....
```

cbuescher replied:
Oh, I see what you mean. Yes, the strange thing here is that inner queries can be `null` when they are coming from the parser, thats why we need the setters/constructors to allow for `null`. I think those checks are there to prevent NPEs later in the doXContent, because we can't render them otherwise. I saw we don't do these checks in other builders, maybe we should? Otherwise we can choose to not render out these fields, but this will lead to parser errors later (there we only check the existence of this field name, not that the query contained actually parsers to a valid query).
If we remove the checks here we simply get NPEs when rendering, like we would get currently in other builders like And/ConstantScore etc... Will think about this, if you have thoughts let me know.

javanna replied:
I think we should have these checks in validate. and if we don't want to rely on validate in doXContent have if conditionals there and ignore what's null just to prevent NPEs

cbuescher replied:
The problem with having these checks in validate() is that when request comes through rest layer, null values here are okay here. They are just not okay when rendering the request as XContent in the builder.

cbuescher replied:
Okay so you suggest to prevent NPEs by null checks in doXContent, then fail later in parser. That's okay I think. I just wanted to point out we can't have these null checks in validate() as long as innerParse might return `null` like today. Will post an update on this.

javanna replied:
I see your point ok, then let's just prevent NPEs but let's not throw exception, maybe recreate the query that we got through rest in the first place, is it the one with an empty query? or without the query object completely?

cbuescher replied:
Have a look at my last commit f3847c6, I changed the builder so it renders out a query even if one of the clauses was not set. This will produce queries like: `{
  "boosting" : {
    "positive" : {
      "match_all" : { }
    },
    "negative_boost" : 0.1,
    "boost" : 1.0
  }
}`

which will be rejected by the parser because there we require the user to state both clauses. It's hard to render a query that will parse but will parse to `null`, I think `{}` doesn't work here and just trying to render the fieldname results in XContentBuilder errors for me. 
The core problem is that we need the builders to allow something when they serve as the "intermediate query representation" (namely null clauses) that they shouldn't support in their original use as json-query builders. Given that we eventually get rid of doXContent here I'd be happy to go this way.

javanna replied:
this is fishy. I think part of what we are doing here is make sure we throw the same errors in the same situations. Before our refactoring everything would go through the parsers, thus validation in the parser rules. After the refactoring the parser can be skipped, meaning that we can't have validation that happens only in the parsers and is skipped in the builders, that is why we have the validate method. Having doXContent producing something that the parser won't accept is a no-go. I think we should go over all the refactored parsers and double check validation that happens there and can be moved to the builder for consistency. Makes sense?

cbuescher replied:
@javanna You're right, the problem remains that (for BoostingQuery) from the rest perspective at least at the moment is required to have a `"postive"/"negative"` section in the json, but they can parse to `null`, so we need the query builder and validate to allow that. I now changed the doXContent to render out an empty json object when inner query is `null`. That would be similar to a user sending a query like `{ "boosting" : { "positive" : { }, "negative" : { }, "negative_boost" : 0.1, "boost" : 1.0 } }` via rest. It will be accepted, but will itself return a `null` query (same behaviour as in the parser before the refactoring).

javanna commented:
thanks @cbuescher for looking into this, I left a small comment inline, also went over all the refactored queries and found something that needs to be updated in SimpleQueryStringBuilder#doXContent .

cbuescher commented:
@javanna With the last commit I added utility method that prevents potential NPEs when trying to render XContent for inner queries that have a `null`reference. I also added `Objects.notNull()` checks where we might potentially add a `null`query to one of the query builder collections (e.g. the clause-lists in BoolQB). 
I'm not sure if we already finally decided on this kind of fail-early-checks in setters/adders/constructors, but doing this in validate would require iterating through all clauses in a list. Also we avoid adding null-clauses on the parser side, so if a user of the builder tries this he should immediately get the exception I think. Let me know if you agree or I should change this back.

javanna commented on core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java:
```diff
@@ -144,4 +143,19 @@ protected static Object convertToStringIfBytesRef(Object obj) {
+        if (queryBuilder != null) {
+            queryBuilder.toXContent(xContentBuilder, params);
+        } else {
+            xContentBuilder.startObject();
```
can we add a comment explaining why we do this. I think to mimic what QueryParseContext#parseInnerQueryBuilder accepts, which converts to null, so null goes back to `{}`. Is that right?

javanna commented on core/src/main/java/org/elasticsearch/index/query/AndQueryBuilder.java:
```diff
@@ -42,12 +42,15 @@
+     */
     public AndQueryBuilder(QueryBuilder... filters) {
         for (QueryBuilder filter : filters) {
             this.filters.add(filter);
```
maybe we should go through the add method here to make sure that null values get rejected properly

javanna commented on core/src/main/java/org/elasticsearch/index/query/FQueryFilterBuilder.java:
```diff
@@ -105,6 +104,12 @@ public Query toQuery(QueryParseContext parseContext) throws QueryParsingExceptio
     @Override
+    public QueryValidationException validate() {
+        // nothing to validate
+        return null;
```
thinking about it... we can address it in another PR, but we have to validate the inner query here too :) something we'll have to fix in all the compouned queries

javanna commented:
This looks great @cbuescher . The natural next step is really replacing those null query builder with a no op query builder as we previously discussed. We can do that on a subsequent PR.

I left a couple of minor comments but LGTM besides those.

One more thing, I think you have to rebase and have a look at CommonTermsQuery and SImpleQueryStringQuery that got merged meanwhile and fix a few things there too.

> I'm not sure if we already finally decided on this kind of fail-early-checks in setters/adders/constructors, but doing this in validate would require iterating through all clauses in a list.

I think it makes sense to do it this way, because it's a check especially for java api users, assuming that our own parsers will never set null there.

cbuescher commented:
@javanna did a rebase, addressed your two remarks concerning additional comment and null checks when adding q.b. to collections. I also added the `flags` field to SimpleQueryStringBuilder equals()/hashCode() unless there was a reason why this was ommited, maybe you can double check this makes sense. CommonTermsQ.B. looked good as far as current default checks/validation is concerned.

javanna commented on core/src/main/java/org/elasticsearch/index/query/AbstractQueryBuilder.java:
```diff
@@ -144,4 +143,20 @@ protected static Object convertToStringIfBytesRef(Object obj) {
+        if (queryBuilder != null) {
+            queryBuilder.toXContent(xContentBuilder, params);
+        } else {
+            // we output an empty object, QueryParseContext#parseInnerQueryBuilder will parse this back to `null`value
```
missing whitespace between `null` and value :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/SimpleQueryStringBuilder.java:
```diff
@@ -438,7 +435,7 @@ public boolean doEquals(SimpleQueryStringBuilder other) {
                 && Objects.equals(defaultOperator, other.defaultOperator) && Objects.equals(queryText, other.queryText)
                 && Objects.equals(queryName, other.queryName) && Objects.equals(minimumShouldMatch, other.minimumShouldMatch)
-                && Objects.equals(settings, other.settings);
+                && Objects.equals(settings, other.settings) && (flags == other.flags);
```
good catch!!!! not sure why our tests didn't uncover this, will dig.

cbuescher replied:
What isn't equal isn't equal. Hash and equals are consistent, hard for any test to see if its okay or not unless we flip every allowed field and check that afterwards things are not equal, I guess.

javanna commented on core/src/main/java/org/elasticsearch/index/query/DisMaxQueryBuilder.java:
```diff
@@ -169,7 +169,7 @@ public DisMaxQueryBuilder readFrom(StreamInput in) throws IOException {
+        out.writeNamedWriteableList(queries);
         out.writeFloat(tieBreaker);
         out.writeOptionalString(queryName);
         out.writeFloat(boost);
```
don't we have to add a null check to the add method here too?

cbuescher replied:
yes, done.

javanna commented on core/src/main/java/org/elasticsearch/index/query/ConstantScoreQueryBuilder.java:
```diff
@@ -79,8 +79,7 @@ public float boost() {
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
```
missing null check in the constructor?

cbuescher replied:
No, inner query builder can parse to null currently.

javanna replied:
gotcha the subtle difference is that here the parser passes in null....alright.

javanna commented on core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java:
```diff
@@ -120,17 +120,9 @@ public float boost() {
+        doXContentInnerBuilder(builder, "negative", negativeQuery, params);
         builder.field("negative_boost", negativeBoost);
         builder.field("boost", boost);
         builder.endObject();
```
should we add null checks in negative and positive setter?

cbuescher replied:
No, same as above.

javanna replied:
gotcha

javanna commented on core/src/main/java/org/elasticsearch/index/query/QueryFilterBuilder.java:
```diff
@@ -60,8 +60,7 @@ public QueryBuilder innerQuery() {
-        builder.field(NAME);
-        queryBuilder.toXContent(builder, params);
+        doXContentInnerBuilder(builder, NAME, queryBuilder, params);
     }
```
missing null check in constructor?

cbuescher replied:
No, same as above.

javanna replied:
gotcha :)

javanna commented on core/src/main/java/org/elasticsearch/index/query/FQueryFilterBuilder.java:
```diff
@@ -79,8 +79,7 @@ public String queryName() {
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
```
missing null check in constructor?

cbuescher replied:
No, same as above.

javanna commented on core/src/main/java/org/elasticsearch/index/query/NotQueryBuilder.java:
```diff
@@ -70,8 +70,7 @@ public String queryName() {
     @Override
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
```
missing null check in constructor?

cbuescher replied:
No, same as above.

javanna commented:
I did another review and found a few places where I think we should add null checks, so that we make sure we never have null inner queries anywhere. again, it is just a check for java api users, cause we expect our parsers to skip null queries, which has a special meaning at the moment.

cbuescher commented:
Added two fixes, the constructor and setter changes can only happen if we change innerParserXY() to always return a QueryBuilder which I hope will happen with the PR we discussed. As long as that we need to allow it. We could add null-checks like when adding to the collections around all setter invocations in the fromXContent methods, but I think this would complicate things even more, given that we plan to change it.

javanna commented:
alright, thanks @cbuescher LGTM
